controller/config/config.go

```diff
@@ -680,50 +681,62 @@ func isSelfSigned(cert *x509.Certificate) (bool, error) {
 }
 
 func generateDefaultSpiffeId(id identity.Identity) (*url.URL, error) {
-	rawCerts := id.Cert().Certificate
-	certs := make([]*x509.Certificate, len(rawCerts))
+	log := pfxlog.Logger()
+	// iterate over the slice of raw server cert chains and return a spiffeId for the first completed chain
+	for _, tlsCert := range id.ServerCert() {
 
-	for i := range rawCerts {
-		cert, err := x509.ParseCertificate(rawCerts[i])
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse certificate at index [%d]: %w", i, err)
+		// initialize a slice of extra certs to pass to GetChain to accompany the leaf for building the chain
+		extraCerts:= make([]*x509.Certificate, len(tlsCert.Certificate))
+
+		// for each raw cert in the chain, parse it and add it to the extraCerts slice
+		for i, cert := range tlsCert.Certificate {
+			parsed, err := x509.ParseCertificate(cert)
+			if err != nil {
+				log.Debugf("failed to parse certificate at index [%d]: %v", i, err)
+				continue
+			}
+			extraCerts[i] = parsed
 		}
-		certs[i] = cert
-	}
 
-	chain := id.CaPool().GetChain(id.Cert().Leaf, certs...)
+		chain := id.CaPool().GetChain(tlsCert.Leaf, extraCerts...)
 
-	// chain is 0 or 1, no root possible
-	if len(chain) <= 1 {
-		return nil, fmt.Errorf("error generating default trust domain from root CA: no root CA detected after chain assembly from the root identity server cert and ca bundle")
-	}
+		// chain is 0 or 1, no root possible
+		if len(chain) <= 1 {
+			log.Debugf("could not generate default trust domain: identity.server_cert contains %d certificates, need at least 2", len(chain))
+			continue
+		}
 
-	candidateRoot := chain[len(chain)-1]
+		candidateRoot := chain[len(chain)-1]
 
-	if candidateRoot == nil {
-		return nil, fmt.Errorf("encountered nil candidate root ca during default trust domain generation")
-	}
+		if candidateRoot == nil {
+			log.Debug("encountered nil candidate root ca during default trust domain generation")
+			continue
+		}
 
-	if !candidateRoot.IsCA {
-		return nil, fmt.Errorf("candidate root CA is not flagged with the x509 CA flag")
-	}
+		if !candidateRoot.IsCA {
+			log.Debug("candidate root CA is not flagged with the x509 CA flag")
+			continue
+		}
 
-	if selfSigned, _ := isSelfSigned(candidateRoot); !selfSigned {
-		return nil, errors.New("candidate root CA is not self signed")
-	}
+		if selfSigned, _ := isSelfSigned(candidateRoot); !selfSigned {
+			log.Debug("candidate root CA is not self signed")
+			continue
+		}
 
-	rawHash := sha1.Sum(candidateRoot.Raw)
+		rawHash := sha1.Sum(candidateRoot.Raw)
 
-	fingerprint := fmt.Sprintf("%x", rawHash)
-	idStr := "spiffe://" + fingerprint
+		fingerprint := fmt.Sprintf("%x", rawHash)
+		idStr := "spiffe://" + fingerprint
 
-	spiffeId, err := url.Parse(idStr)
+		spiffeId, err := url.Parse(idStr)
 
-	if err != nil {
-		return nil, fmt.Errorf("could not parse generated SPIFFE id [%s] as a URI: %w", idStr, err)
-	}
+		if err != nil {
```

Comment:
What is w/ all the whitespace change?

Reply:
I wouldn't expect formatting changes due to Go's opinionated formatting and that Goland auto formats most things.

Reply:
For some reason, the `Identity.ServerCert()` type is a slice of cert byte slices, whereas `Identity.Cert()` is a slice of bytes, so I needed to indent one additional iteration or limit to only the first chain. 

I went with iterating, though it's admittedly unclear to me when there could be more than one. This prompted me to change some returns to continue to allow for a subsequent iteration to find a complete chain and generate the default spiffe id.

Reply:
More precisely,

```go
type Identity interface {
	Cert() *tls.Certificate
	ServerCert() []*tls.Certificate
        ...
}
```

I was referring to the fact that Cert() is a point to a cert, whereas ServerCert() is a slice of pointers, so I need one more indent in order to iterate over the slice.

Reply:
If the first chain doesn't resolve to self-signed, should we return with an error instead of iterating then? We should use an alt chain for the default spiffy ID, right?

Reply:
I am not sure that this code would defend against a full root chain being provided in altServerCerts that also has SPIFFE IDs.

Reply:
In case of conflation, this is the portion of the code that generates a default id as a last resort, not the part that looks for SPIFFE IDs in URI SANs.

Reply:
That is correct. However, I still see an issue if an alt server cert + root is provided.

It will be to the wrong root. We would want the root of the network's PKI not some other random PKI. Unfortunately the identity interface doesn't distinguish between the piles of certs.

Reply:
If the first chain doesn't resolve to self-signed, should we return with an error instead of iterating then? We should NOT use an alt chain for the default spiffy ID, right?